Metric API v0.4 specification (#581)
open-telemetry/opentelemetry-specification: specification/metrics/api.md

### Metric Instruments

A _metric instrument_ is a device for capturing raw measurements in

> the API.  There are six kinds of instrument, each with a dedicated

Author: I removed "six" everywhere.

purpose.  The API purposefully avoids optional features that change
the semantic interpretation of an instrument; the API instead prefers
instruments that support a single method with fixed interpretation.

All measurements captured by the API are associated with
instrument used to make the measurement, thus giving the measurement its semantic properties.
Instruments are created and defined through calls to a `Meter` API,
which is the user-facing entry point to the SDK.

Instruments are classified in several ways that distinguish them from
one another.

1. Synchronicity: A synchronous instrument is called by the user in an OpenTelemetry [Context](../context/context.md). An asynchronous instrument is called by the SDK once per collection interval, lacking a Context.

> 2. Additivity: An additive instrument is used to capture information about a sum of values.  A non-additive instrument is used to capture information about individual values.

Author: This is mostly the same question @MrAlias asked. I'm having trouble finding a good explanation, I guess.

> 2. Additivity: An additive instrument is used to capture information about a sum of values.  A non-additive instrument is used to capture information about individual values.

Reviewer: Random thought: Since we don't expose this as a property on the API side, maybe just removing the very detailed explanation of additivity and monotonicity for the instruments makes sense. Keeping something simpler like what we had before that user intend to see a `Sum` and if the sum always goes up or goes up and down is enough. I may oversimplifying things but want some options.

> 2. Additivity: An additive instrument is used to capture information about a sum of values.  A non-additive instrument is used to capture information about individual values.

Reviewer: I know that we talked a lot about that for us to understand deeper every detail, but when it comes to the user of the API that may not be necessary.

> 2. Additivity: An additive instrument is used to capture information about a sum of values.  A non-additive instrument is used to capture information about individual values.

Author: See 091119014240eab08990936d6b4faf6ac530039b.

> 2. Additivity: An additive instrument is used to capture information about a sum of values.  A non-additive instrument is used to capture information about individual values.

Reviewer: I think that clarifies a lot of things. But my "random thought" still stays. Do we give too many details in the API documentation?

> 2. Additivity: An additive instrument is used to capture information about a sum of values.  A non-additive instrument is used to capture information about individual values.

Author: I keep trying to ensure that the API is described in terms that we know what the API calls mean and not what they do. If we say that the additive instruments are only additive when used with the default aggregation, I think it means we've defined what they do and not what they mean. It's harder to define semantics w/o referring to behavior.

I'm inclined to say that this document is meant for implementors, not users, and that this sort of detail matters. I recognize that this kicks something down the road--eventually some descriptive guide will have to be created for users, and it can try to describe the instruments in a different way, but I think it's useful to have a term that distinguishes data where only the sum is considered valuable information, compared with data where the median or arithmetic mean is considered valuable information. I think we haven't found a better term for this, yet.

3. Monotonicity: An additive instrument can be monotonic, when the sequence of captured sums is non-decreasing.  Monotonic instruments are useful for monitoring rate information.

The metric instruments names are shown below along with whether they
are synchronous, additive, and/or monotonic.

| Name | Synchronous | Additive | Monotonic |
| ---- | ----------- | -------- | --------- |
| Counter           | Yes | Yes | Yes |
| UpDownCounter     | Yes | Yes | No  |
| ValueRecorder     | Yes | No  | No  |
| SumObserver       | No  | Yes | Yes |
| UpDownSumObserver | No  | Yes | No  |
| ValueObserver     | No  | No  | No  |

The synchronous instruments are useful for measurements that are
gathered in a tracing context.  The asynchronous instruments are
useful when measurements are expensive, therefore should be gathered
periodically.  Read more [characteristics of synchronous and
asynchronous instruments](#synchronous-and-asynchronous-instruments-compared) below.

> The additive instruments are useful in situations where we are only

Author: I added a section defining measurements and the term "additive", which I think helps. See 091119014240eab08990936d6b4faf6ac530039b.

interested in the sum, as this enables significant optimizations in
the export pipeline.  The synchronous and asynchronous additive
instruments have a significant difference: synchronous instruments are
used to capture changes in a sum, whereas asynchronous instruments are
used to capture sums directly.  Read more [characteristics of additive
instruments](#additive-and-non-additive-instruments-compared) below.

> The monotonic instruments are significant because they support rate

Author: I added more to state that monotonic instruments are additive. I was trying to keep this material introductory, so the topic is covered later in more depth.

covered in the [user-level API specification](api-user.md).

### Labels

_Label_ is the term used to refer to a key-value attribute associated
with a metric event, similar to a [Span
attribute](../trace/api.md#span) in the tracing API.  Each label
categorizes the metric event, allowing events to be filtered and
grouped for analysis.

Each of the instrument calling conventions detailed in the [user-level
API specification](api-user.md) accepts a set of labels as an
argument.  The set of labels is defined as a unique mapping from key to value.
Typically, labels are passed to the API in the form of a list of
key:values, in which case the specification dictates that duplicate
entries for a key are resolved by taking the last value to appear in the list.
This is known as "last-value wins".

Measurements by a synchronous instrument are commonly combined with
other measurements having exactly the same label set, which enables
significant optimizations.  Read more about [combining measurements
through aggregation](#aggregations) below.

Asynchronous instruments are permitted to observe at most one value
per distinct label set, per callback invocation.  Read more about

> [current label sets for asynchronous instruments](#asynchronous-observations-form-a-current-set) below.

Author: I deleted this paragraph. The material is covered later.

### Meter Interface

The API defines a `Meter` interface.  This interface consists of a set
of instrument constructors, and a facility for capturing batches of
measurements in a semantically atomic way.

There is a global `Meter` instance available for use that facilitates
automatic instrumentation for third-party code.  Use of this instance
allows code to statically initialize its metric instruments, without
explicit dependency injection.  The global `Meter` instance acts as a
no-op implementation until the application explicitly initializes a

> global `Meter` by installing an SDK.  Note that it is not necessary to

Author: > ... installing an SDK either explicitly, through a service provider interface, or other language-specific support.

The synchronous instruments are useful for measurements that are
gathered in a tracing context.  The asynchronous instruments are
useful when measurements are expensive, therefore should be gathered
periodically.  Read more [characteristics of synchronous and
asynchronous instruments](TODO) below.

The additive instruments are useful in situations where we are only
interested in the sum, as this enables significant optimizations in
the export pipeline.  The synchronous and asynchronous additive
instruments have a significant difference: synchronous instruments are
used to capture changes in a sum, whereas asynchronous instruments are
used to capture sums directly.  Read more [characteristics of additive
instruments](TODO) below.

The monotonic instruments are significant because they support rate
calculations.  Read more information about [choosing metric
instruments](TODO) below.

An _instrument definition_ describes several properties of the
instrument, including its name and its kind.  The other properties of
a metric instrument are optional, including a description and the unit
of measurement.  An instrument definition is associated with the

> events that it produces.

Reviewer 2: Above, a "metric event" is defined to be the thing when the value is captured, but here "event" is being used to refer to the output of an instrument. I get what the meaning is here, but it would be nice if we had a word to describe the output of an instrument that didn't collide with the name ascribed to the capturing of the data by the instrument. 

As a follow-on, I'm actually not sure what it means for an instrument to produce events, especially from the API perspective. I wonder if we should just elide this sentence from the API specification, as this is more of an SDK concern.

Details about calling conventions for each kind of instrument are
covered in the [user-level API specification](api-user.md).

### Labels

_Label_ is the term used to refer to a key-value attribute associated
with a metric event, similar to a [Span
attribute](../trace/api.md#span) in the tracing API.  Each label
categorizes the metric event, allowing events to be filtered and
grouped for analysis.

Each of the instrument calling conventions detailed in the [user-level

> API specification](api-user.md) accepts a list of labels as an

Reviewer 2: Is it mandatory that labels are passed as a `list`? This seems like it could definitely vary by language. Can we just specify that a `Set` of labels needs to be accepted and add a recommendation of "last value wins" when optimization are made to have them passed as a list?

argument, which are used to compute a label set.  The set of labels is
defined as a unique mapping from key to value, but since they are
passed in as a list, the specification dictates that duplicate keys are

> empowersresolved by taking the last value to appear in the list.  This is known

Reviewer 2: Something got munged on this line : `empowersresolved`

as "last-value wins".

Measurements by a synchronous instrument are commonly combined with
other measurements having exactly the same label set, which enables
significant optimizations.  Read more about [combining measurements
through aggregation](TODO) below.

Asynchronous instruments are permitted to observe at most one value
per distinct label set, per callback invocation.  Read more about
[current label sets for asynchronous instruments](TODO) below.

### Meter Interface

> To produce measurements using an instrument, you need an SDK that implements

Reviewer 2: This is an odd sentence to have in an API spec. I might just say "The API defines a `Meter` interface."

the `Meter` API.  This interface consists of a set of instrument constructors,
and a facility for capturing batches of measurements in a semantically atomic
way.

There is a global `Meter` instance available for use that facilitates
automatic instrumentation for third-party code.  Use of this instance
allows code to statically initialize its metric instruments, without
explicit dependency injection.  The global `Meter` instance acts as a
no-op implementation until the application explicitly initializes a
global `Meter` by installing an SDK.  Note that it is not necessary to
use the global instance: multiple instances of the OpenTelemetry SDK
may run simultaneously.

As an obligatory step, the API requires the caller to provide the name
of the instrumenting library (optionally, the version) when obtaining
a `Meter` implementation.  The library name is meant to be used for
identifying instrumentation produced from that library, for such
purposes as disabling instrumentation, configuring aggregation, and
applying sampling policies.  See the specification on [obtaining a

> Tracer](../trace/api.md#obtaining-a-tracer) for more details.

Reviewer 3: Does this mean we'll have MeterProvider (similar to TracerProvider), and there will be a global default MeterProvider, and users can create more MeterProviders if they wish? Like everything from TracerProvider, but named MeterProvider?

> Tracer](../trace/api.md#obtaining-a-tracer) for more details.

Reviewer: Yes

Details about installing an SDK and obtaining a named `Meter` are
covered in the [SDK-level API specification](api-meter.md).

### Aggregations

_Aggregation_ refers to the process of combining multiple measurements
into exact or estimated statistics about the metric

> events that took place during an interval time, during program

Author: I added a section defining measurements and the term "additive". See 091119014240eab08990936d6b4faf6ac530039b.

execution.

Each instrument specifies a default aggregation that is suited to the
semantics of the instrument, that serves to explain its properties and
give users an understanding of how it is meant to be used.
Instruments, in the absence of any configuration override, can be
expected to deliver a useful, economical aggregation out of the box.

The additive instruments (`Counter`, `UpDownCounter`, `SumObserver`,

> `UpDownSumObserver`) use Sum aggregation, by default.  Details about

Reviewer 2: ```suggestion
`UpDownSumObserver`) use a Sum aggregation by default.  Details about
```

computing a Sum aggregation vary, but from the user's perspective this

> means they will be able to monitor the sum of values entered.  The

Reviewer 2: ```suggestion
means they will be able to monitor the sum of the values recorded.  The
```

distinction between synchronous and asynchronous instruments is
crucial to specifying how exporters work, a topic that is covered in
the [SDK specification](TODO).

The non-additive instruments (`ValueRecorder`, `ValueObserver`) use

> MinMaxSumCount aggregation, by default.  This aggregation keeps track

Reviewer 2: ```suggestion
a MinMaxSumCount aggregation, by default.  This aggregation keeps track
```

of the minimum value, the maximum value, the sum of values, and the
count of values.  These four values support monitoring the range of
values, the rate of events, and the average event value.  

Other standard aggregations are available, especially for non-additive
instruments, where we are generally interested in a variety of
different summaries, such as histograms, quantile summaries,
cardinality estimates, and other kinds of sketch data structure.

> The default OpenTelemetry SDK implements a [Views API](TODO), which

Reviewer 2: Is it ok to mention this before the OTEP has even been approved?

### Metric Event Format

Metric events have the same logical representation, regardless of
instrument kind.  Metric events captured through any instrument
consist of:

- timestamp (implicit)
- instrument definition (name, kind, description, unit of measure)
- label set (keys and values)
- value (signed integer or floating point number)
- [resources](../resource/sdk.md) associated with the SDK at startup.

Synchronous events have one additional property, the distributed
[Context](../context/context.md) (Span context, Correlation context)

> that was active at the time.

Reviewer: I like this way of specifying the association with "distributed context". Maybe use the same terminology in the previous comment as well. See my comment above.

Author: Done.

## Instrument properties

Because the API is separated from the SDK, the implementation
ultimately determines how metric events are handled.  Therefore, the
choice of instrument should be guided by semantics and the intended
interpretation.  The semantics of the individual instruments is
defined by several properties, detailed here, to assist with
instrument selection.

### Synchronous and asynchronous instruments compared

> Synchronous instruments are called in a request context, meaning they

Reviewer: Confusing that in the doc I saw:
* tracing context.
* distributed context.
* request context.

We should try to be consistent :)

Author: OK. I've updated the document using "distributed [Context](../context/context.md) (i.e., Span context, Correlation context)" consistently, with links everywhere to be clear. 

I've used "during a request" or "inside a request" to describe the situation in which synchronous events happen. This can probably still be improved.

have an associated tracing context and distributed
correlation values.  Multiple metric events may occur for a
synchronous instrument within a given collection interval.

Asynchronous instruments are reported by a callback, once per
collection interval, and lack request context.  They are permitted to
report only one value per distinct label set per period.  If the
application observes multiple values for the same label set, in a

> single callback, the last value is the only value kept.

Reviewer: There is a previous comment about this. I think you should update there as well.

nit: I feel that some of these informations are duplicated, not very important but you may take a look.

Author: I removed the earlier statement, it was less thorough.

### Metric Event Format

Metric events have the same logical representation, regardless of
instrument kind.  Metric events captured through any instrument
consist of:

- timestamp (implicit)
- instrument definition (name, kind, description, unit of measure)
- label set (keys and values)
- value (signed integer or floating point number)
- [resources](../resource/sdk.md) associated with the SDK at startup.

Synchronous events have one additional property, the distributed
[Context](../context/context.md) (i.e., Span context, Correlation context)

> that was active at the time.

Reviewer 3: Where and how is the context values used? Are these context used as additional labels? I vaguely remember reading in some spec that none of the standard aggregators use this.

Author: Ah, this was omitted. Thanks! I've added:

```
### Association with distributed context

Synchronous measurements are implicitly associated with the
distributed [Context](../context/context.md) at runtime, which may
include a Span context and Correlation values.  The Metric SDK may use
this information in many ways, but one feature is of particular
interest in OpenTelemetry.

#### Correlation context into metric labels

Correlation context is supported in OpenTelemetry as a means for
labels to propagate from one process to another in a distributed
computation.  Sometimes it is useful to aggregate metric data using
distributed correlation values as metric labels.

The use of correlation context must be explicitly configured, using
the [Views API (WIP)](https://github.com/open-telemetry/oteps/pull/89)
to select specific key correlation values that should be applied as
labels.  The default SDK will not automatically use correlation
context labels in the export pipeline, since using correlation labels
can be a significant expense.

Configuring views for applying Correlation context labels is a [work in
progress](https://github.com/open-telemetry/oteps/pull/89).

```

report only one value per distinct label set per period.  If the
application observes multiple values for the same label set, in a
single callback, the last value is the only value kept.

To ensure that the definition of last value is consistent across
asynchronous instruments, the timestamp associated with asynchronous
events is fixed to the timestamp at the end of the interval in which
it was computed.  All asynchronous events are timestamped with the end
of the interval, which is the moment they become the last value
corresponding to the instrument and label set.  (For this reasons,
SDKs SHOULD run asynchronous instrument callbacks near the end of the
collection interval.)

### Additive and non-additive instruments compared

Additive instruments are used to capture information about a sum,
where, by definition, only the sum is of interest.  Individual events
are considered not meaningful for these instruments, the event count
is not computed.  This means, for example, that two `Counter` events
`Add(N)` and `Add(M)` are equivalent to one `Counter` event `Add(N +
M)`.  This is the case because `Counter` is synchronous, and

> synchronous additive instruments are used to capture changes to a sum.

Author: Let me know 091119014240eab08990936d6b4faf6ac530039b helps.

Asynchronous, additive instruments (e.g., `SumObserver`) are used to
capture sums directly.  This means, for example, that in any sequence
of `SumObserver` observations for a given instrument and label set,
the Last Value defines the sum of the instrument.

In both synchronous and asynchronous cases, the additive instruments
are inexpensively aggregated into a single number per collection interval
without loss of information.  This property makes additive instruments
higher performance, in general, than non-additive instruments.

Non-additive instruments use a relatively inexpensive aggregation
method default (MinMaxSumCount), but still more expensive than the
default for additive instruments (Sum).  Unlike additive instruments,
where only the sum is of interest by definition, non-additive

> instruments can be configured with even more expensive aggregators.

Reviewer 3: this sentence gives the feeling that we won't allow configuring a different aggregator for additive instruments. Is that intended?

Author: Not intended. I added a section defining measurements and the term "additive". See 091119014240eab08990936d6b4faf6ac530039b.

### Monotonic and non-monotonic instruments compared

Monotonicity applies only to additive instruments.  `Counter` and
`SumObserver` instruments are defined as monotonic because the sum
captured by either instrument is non-decreasing.  The `UpDown-`
varations of these two instruments are non-monotonic, meaning the sum
can increase, decrease, or remain constant without any guarantees.

Monotonic instruments are commonly used to capture information about a
sum, where the sum itself is less relevant than the rate expressed by
the sum's change over time.  The Monotonic property is defined by this
API to refer to a non-decreasing sum.  Non-increasing sums are not

> considered a feature in the Metric API.

Author: I removed the "less relevant" part and just wrote "a sum that is meant to be monitored as a rate" to avoid the confusion.

### Function names

Each instrument supports a single function, named to help convey the
instrument's semantics.

Synchronous additive instruments support an `Add()` function,
signifying that they add to a sum and do not directly capture a sum.

Synchronous non-additive instruments support a `Record()` function,
signifying that they capture individual events, not only a sum.

Asynchronous instruments all support an `Observe()` function,
signifying that they capture only one value per measurement interval.

## The instruments

### Counter

`Counter` is the most common synchronous instrument.  This instrument
supports an `Add(increment)` function for reporting a sum, and is
restricted to non-negative increments.  The default aggregation is
`Sum`, as for any additive instrument.

Example uses for `Counter`:

- count the number of bytes received
- count the number of accounts created
- count the number of checkpoints run

> - count the number of 5xx errors.

Author: Done.

These example instruments would be useful for monitoring the rate of
any of these quantities.  In these situations, it is usually more
convenient to report by how much a sum changes, as it happens, than to
calculate and report the sum on every measurement.

### UpDownCounter

`UpDownCounter` is similar to `Counter` except that `Add(increment)`
supports negative increments.  This makes `UpDownCounter` not useful
for computing a rate aggregation.  It aggregates a `Sum`, only the sum
is non-monotonic.  It is generally useful for capturing changes in an
amount of resources used, or any quantity that rises and falls during a
request.

Example uses for `UpDownCounter`:

- count memory in use by instrumenting `new` and `delete`
- count queue size by instrumenting `enqueue` and `dequeue`

> - count semaphore `up` and `down` operations.

Author: Done.

These example instruments would be useful for monitoring resource
levels across a group of processes.

### ValueRecorder

`ValueRecorder` is a non-additive synchronous instrument useful for
recording any non-additive number, positive or negative.  Values
captured by a `Record(value)` are treated as individual events
belonging to a distribution that is being summarized.  `ValueRecorder`
should be chosen either when capturing measurements that do not
contribute meaningfully to a sum, or when capturing numbers that are
additive in nature, but where the distribution of individual
increments is considered interesting.

One of the most common uses for `ValueRecorder` is to capture latency
measurements.  Latency measurements are not additive in the sense that
there is little need to know the latency-sum of all processed
requests.  We use a `ValueRecorder` instrument to capture latency
measurements typically because we are interested in knowing mean,
median, and other summary statistics about individual events.

The default aggregation for `ValueRecorder` computes the minimum and
maximum values, the sum of event values, and the count of events,

> allowing the rate, the mean, and and range of input values to be

Reviewer 3: nit: and and - twice.

Author: Fixed.

monitored.

Example uses for `ValueRecorder` that are non-additive:

- capture any kind of timing information
- capture the acceleration experienced by a pilot
- capture nozzle pressure of a fuel injector
- capture the velocity of a MIDI key-press.

Example _additive_ uses of `ValueRecorder` capture measurements that
are additive, but where we may have an interest in the distribution of
values and not only the sum:

- capture a request size
- capture an account balance
- capture a queue length
- capture a number of board feet of lumber.

These examples show that although they are additive in nature,
choosing `ValueRecorder` as opposed to `Counter` or `UpDownCounter`
implies an interest in more than the sum.  If you did not care to
collect information about the distribution, you would have chosen one
of the additive instruments instead.  Using `ValueRecorder` makes
sense for capturing distributions that are likely to be important in
an observability setting.

Use these with caution because they naturally cost more than the use
of additive measurements.

### SumObserver

> `SumObserver` is the asynchronous instrument corresponding to

Reviewer 3: Writing down an example I used for my visualization. Feel free to discard/add a similar example to avoid any potential confusions.

If a user has both bytes_sent and total_bytes_sent available (for whatever reason), choosing SumObserver means he would do sumObserver.Observe(total_bytes_sent), and chosing Counter means he would do counter.Add(bytes_sent).

If the user only has bytes_sent, and is not interested in keeping total_bytes_sent himself, then use Counter, and OT metric keep total sum.
If the user only has total_bytes_sent, then use SumObserver, and always report the total_bytes_sent.

Author: Yes, you've got it. I haven't added this, it feels like a more long-form example of the sort that Bogdan is trying to get me to leave out.

`ValueObserver` reports a current sum (equal to the max and the min)
and a count equal to 1.  When there is no aggregation, these results
are equivalent.

It may seem pointless to define a default aggregation when there is
exactly one data point.  The default aggregation is specified to apply
when performing spatial aggregation, meaning to combine measurements
across label sets or in a distributed setting.  Although a
`ValueObserver` observes one value per collection interval, the
default aggregation specifies how it will be aggregated with other
values, absent any other configuration.

Therefore, considering the choice between `ValueObserver` and
`UpDownSumObserver`, the recommendation is to choose the instrument
with the more-appropriate default aggregation.  If you are observing a
queue size across a group of machines and the only thing you want to
know is the aggregate queue size, use `SumObserver` because it
produces a sum, not a distribution.  If you are observing a queue size
across a group of machines and you are interested in knowing the
distribution of queue sizes across those machines, use
`ValueObserver`.

> ### Interpretation

Reviewer: I think everything below this point can go in an advance section in a different doc (maybe different PR) if we want to make faster progress with the API, everything above is very good to start ensuring consistency for all the languages, up to you (PS: I did not review below this, since it seems just more details that can be added later).

Author: :shrug:

How are the instruments fundamentally different, and why are there
only three?  Why not one instrument?  Why not ten?

As we have seen, the instruments are categorized as to whether
they are synchronous, additive, and/or and monotonic.  This approach
gives each of the instruments unique semantics, in ways that

> meaningfully improve the performance and interpreation of metric

Reviewer 3: interpreation  -> interpretation

Author: Fixed.

the user, although for some exporters it cannot be avoided.  The
potential for unbounded memory growth comes from "momentary-use"
labels, labels which are used briefly and then are not used again.
If the SDK is required to retain memory of every combination of
instrument and label set it has ever seen, long-lived memory can
become a problem.

Nevertheless, exporters may be forced to allocate long-lived memory to
perform their function, particularly with respect to the additive
instruments.  Metric exposition formats commonly have two ways of
exposing additive instrument data: in terms of the change in a sum, or
in terms of the sum itself.  (The terms "delta" and "cumulative" are
used to describe these two approaches in an export pipeline, we avoid
their use to describe the additive instruments.)

There are two cases where an exporter must retain memory in order to
operate correctly.

1. An exporter that exports sums but receives changes in the sum
2. An exporter that exports changes in a sum but receives the sum.

In both of these cases, the exporter must maintain memory of the last
value of the sum, to convert between the two representations of

> additive data when both forms of input are present.

Reviewer 3: While I appreciate the detailed description, I think this should be part of SDK spec, not really required to use API.

Author: True. I removed this section, it'll go in the SDK spec. Thanks.

instrument selection.

### Synchronous and asynchronous instruments compared

Synchronous instruments are called in a request context, meaning they
potentially have an associated tracing context and distributed
correlation values.  Multiple metric events may occur for a
synchronous instrument within a given collection interval.

Asynchronous instruments are reported by a callback, once per
collection interval, and lack request context.  They are permitted to
report only one value per distinct label set per period.  If the
application observes multiple values for the same label set, in a
single callback, the last value "wins".

To ensure that the definition of last value is consistent across
asynchronous instruments, the timestamp associated with asynchronous
events is fixed to the timestamp at the end of the interval in which
it was computed.  All asynchronous events are timestamped with the end
of the interval, which is the moment they become the last value
corresponding to the instrument and label set.  (For this reasons,
SDKs SHOULD run asynchronous instrument callbacks near the end of the
collection interval.)

### Additive and non-additive instruments compared

> Additive instruments are used to capture information about a sum,

Author: I agree there's something missing here, but I am not sure how to state it. I think of the composition _of the sum_ as being still important, in terms of the contribution from different label sets making up a whole sum. The statement is about the individual events, particularly the count (and rate) of them. I'm looking for suggestions.

> Additive instruments are used to capture information about a sum,

Author: I agree there's something missing here, but I am not sure how to state it. I think of the composition _of the sum_ as being still important, in terms of the contribution from different label sets making up a whole sum. The statement is about the individual events, particularly the count (and rate) of them. I'm looking for suggestions.

@bogdandrutu Has often tried to explain these instruments by admitting that some kind of pre-aggregation is happening before the user calls the API. So the user has applied addition to aggregate some numbers before reporting them through an additive instrument. This explains why the events are meaningless.

> Additive instruments are used to capture information about a sum,

Reviewer 4: @jmacd: I'm running a bit low on time at the moment, but I'll take another pass at this section tomorrow and see if I can make more concrete suggestions on the phrasing.

FWIW, I think keeping the additive and monotonic descriptions is a useful approach. It help explain in more concrete terms the expectations about what the instrumenter is implying and what the aggregator (possibly even the exporter) can expect.

> Additive instruments are used to capture information about a sum,

Author: This also defines "capture" and tries to answer your question. See 091119014240eab08990936d6b4faf6ac530039b.

where, by definition, only the sum is of interest.  Individual events

> are considered not meaningful for these instruments, the event count

Author: I want to emphasize what you wrote above,

> makes it clear in the distinction from capturing the sum directly, like in the async case

This seems to be the challenge. I'm looking for suggestions.

> are considered not meaningful for these instruments, the event count

Author: I added a section defining measurements and the term "additive", then removed this paragraph. See 091119014240eab08990936d6b4faf6ac530039b.

directly.  Use a `SumObserver` to capture any value that starts at
zero and rises throughout the process lifetime and never falls.

Example uses for `SumObserver`.
- capture process user/system CPU seconds
- capture the number of cache misses.

A `SumObserver` is a good choice in situations where a measurement is
expensive to compute, such that it would be wasteful to compute on
every request.  For example, a system call is needed to capture
process CPU usage, therefore it should be done periodically, not on
each request.  A `SumObserver` is also a good choice in situations
where it would be impractical or wasteful to instrument individual
changes that comprise a sum.  For example, even though the number of
cache misses is a sum of individual cache-miss events, it would be too
expensive to synchronously capture each event using a `Counter`.

### UpDownSumObserver

`UpDownSumObserver` is the asynchronous instrument corresponding to
`UpDownCounter`, used to capture a non-monotonic count with
`Observe(sum)`.  "Sum" appears in the name to remind users that it is
used to capture sums directly.  Use a `UpDownSumObserver` to capture

> any value that starts at zero and rises or falls throughout the

Author: I wanted to sweep this issue under the rug. 

I'm not sure we should call this a "Sum" instrument if it doesn't start at zero. We could be talking about cumulative instruments which are only useful for exporting a _derivative_.  If the derivative is all that's important, then I think you're right. If the sum matters, I feel that zero is an implied starting point.

I remember some earlier discussions about how non-negative-rate instruments could be defined with zero as the starting point or not. Looking for input from @bogdandrutu

> any value that starts at zero and rises or falls throughout the

Reviewer 4: I must have missed the previous discussion, sorry for reintroducing. I'm fine with sweeping under the rug. It is such an edge case it is not worth the effort.

the SDK control over how often they are evaluated.

Example uses for `ValueObserver`:
- capture CPU fan speed
- capture CPU temperature.

Note that these examples use non-additive measurements.  In the
`ValueRecorder` case above, example uses were given for capturing
synchronous cumulative measurements in a request context (e.g.,
current queue size seen by a request).  In the asynchronous case,
however, how should users decide whether to use `ValueObserver` as
opposed to `UpDownSumObserver`?

Consider how to report the size of a queue asynchronously.  Both
`ValueObserver` and `UpDownSumObserver` logically apply in this case.
Asynchronous instruments capture only one measurement per interval, so
in this example the `SumObserver` reports a current sum, while the
`ValueObserver` reports a current sum (equal to the max and the min)
and a count equal to 1.  When there is no aggregation, these results
are equivalent.

The recommendation is to choose the instrument with the
more-appropriate default aggregation.  If you are observing a queue
size across a group of machines and the only thing you want to know is
the aggregate queue size, use `SumObserver`.  If you are observing a

> queue size across a group of machines and you are interested in

Author: I added a paragraph.

```
It may seem pointless to define a default aggregation when there is
exactly one data point.  The default aggregation is specified to apply
when performing spatial aggregation, meaning to combine measurements
across label sets or in a distributed setting.  Although a
`ValueObserver` observes one value per collection interval, the
default aggregation specifies how it will be aggregated with other
values, absent any other configuration.
```

labels, labels which are used briefly and then are not used again.
If the SDK is required to retain memory of every combination of
instrument and label set it has ever seen, long-lived memory can
become a problem.

Nevertheless, exporters may be forced to allocate long-lived memory to
perform their function, particularly with respect to the additive
instruments.  Metric exposition formats commonly have two ways of
exposing additive instrument data: in terms of the change in a sum, or
in terms of the sum itself.  (The terms "delta" and "cumulative" are
used to describe these two approaches in an export pipeline, we avoid
their use to describe the additive instruments.)

There are two cases where an exporter must retain memory in order to
operate correctly.

1. An exporter that exports sums but receives changes in the sum
2. An exporter that exports changes in a sum but receives the sum.

In both of these cases, the exporter must maintain memory of the last
value of the sum, to convert between the two representations of
additive data when both forms of input are present.

An exporter can avoid this memory requirement by supporting exposition

> of additive metric data in both ways, as with OTLP.  A client

Reviewer 4: ```suggestion
of additive metric data in both ways, as with the [OpenTelemetry protocol buffer](https://github.com/open-telemetry/opentelemetry-proto) (OTLP).  A client
```

> configured with OTLP for metrics export does not have a long-lived

Reviewer 4: ```suggestion
configured to export the OTLP for metrics does not have a long-lived
```

Author: Done.

memory requirement, because that protocol supports both forms of
additive data.  Although, the memory requirement then arises in the
downstream system for any exporter that does not support both forms.

It is tempting to consider not supporting one of these forms of
additive metric data, as a way to avoid exporter memory requirements.
However, legacy protocols exist with both forms, so this would not
necessarily help.  Besides, the fact that synchronous additive
instruments accept one, while asynchronous additive instruments
accepts the other, is ultimately meant as a user convenience.  Had
OpenTelemetry not specified that asynchronous additive instruments
accept sums directly, this burden would fall to the user (e.g., since
operating systems report heap usage as a sum, the user would be forced
into a memory requirement in order to report changes in the sum).

### Asynchronous observations form a current set

Asynchronous instrument callbacks are permitted to observe one value

> per instrument, per distinct label set, per callback invokation.  The

Reviewer 4: ```suggestion
per instrument, per distinct label set, per callback invocation.  The
```

> set of values recorded by one callback invokation represent a current

Reviewer 4: ```suggestion
set of values recorded by one callback invocation represent a current
```

Author: Done.

snapshot of the instrument; it is this set of values that defines the
Last Value for the instrument until the next collection interval.

Asynchronous instruments are expected to record an observation for
every label set that it considers "current".  This means that
asynchronous callbacks are expected to observe a value, even when the

> value has not changed since the last callback invokation.  To not

Reviewer 4: ```suggestion
value has not changed since the last callback invocation.  To not
```

Author: Done.

observe a label set implies that a value is no longer current.  The
Last Value becomes undefined, as it is no longer current, when it is
not observed during a collection interval.

The definition of Last Value is possible for asynchronous instruments,
because their collection is coordinated by the SDK and because they
are expected to report all current values.  Another expression of this
property is that an SDK can keep just one collection interval worth of
observations in memory to lookup the current Last Value of any
instrument and label set.  In this way, asynchronous instruments
support querying current values, independent of the duration of a

> collection interval, using data collected at a single point in time.

Reviewer 4: This is better fitting in the `api-user.md` definitions.

Author: Did you have in mind moving part of this paragraph or section into the other document? I can see adding similar content there and leaving this in. Which part were you referring to?

Reviewer 4: I was thinking this entire paragraph, but with the recent decision to merge things to this doc this comment can be disregarded and resolved (sorry, not able to resolve myself).
